### ISSUE TITLE
ARMAC implementation

### ISSUE THREAD

**Issue by StochasticEntropy** [deepmind/open_spiel#441]
Is there a plan to include the ARMAC implentation in open spiel? https://arxiv.org/abs/2008.12234

**Comment by lanctot** [comment]
Hi @StochasticEntropy, I would personally love to see ARMAC in OpenSpiel. Our internal implemention depends on a lot of internal code that we can't open source, so we would have to reimplement it. 

I will bring up this thread with the primary author. There is currently no plan, so I will mark it as a welcome contribution, but there is a chance that I look into this myself. If so, I will use this thread for updates.

**Comment by lanctot** [comment]
Tagging @backpropper who might be interested.

**Comment by lanctot** [comment]
Tagging @jhtschultz too. The more I think about it, the more it might make sense to wait until the integration with Acme (see https://github.com/deepmind/acme/issues/80) comes along because it'd be cool to have a scalable implementation.

**Comment by lanctot** [comment]
Finally tagging @agruslys :)

**Comment by StochasticEntropy** [comment]
If there's interest, I'd also be interested in contributing to a JAX implementation, I'm aslo a JAX newbie though.

**Comment by lanctot** [comment]
@dvstter just curious how your impl is looking? Would you be willing to contribute it to OpenSpile?

**Comment by StochasticEntropy** [comment]
Since this was my request initally, I'd also still be interested in helping to implement this. Either by integrating an already existing implementation, or maybe writing a new implementation if one of the original authors would be interested in helping with/proof reading the code.

**Comment by StochasticEntropy** [comment]
Maybe at some point I will have to get a bit suspicious about this paper, since there was no follow up after it got rejected.

Probably the authors got really fed up, for not being published, despite being from google.

But I liked the idea when I first read it, and I still do ....

**Comment by lanctot** [comment]
@StochasticEntropy. You know that I'm one of the authors, right...?

I'd really like to see ARMAC in OpenSpiel. I've wanted to add it myself, but I just have not been able to get to it because there are too many other things that I have had to do. The internal implementation might be too difficult to port, so I'd rather start fresh with a new one (probably in Jax).

It would be great if you wanted to take this on. I would be more than willing to help you out with the implementation (and recruit some help from the other authors as well) Let me know what you think.

**Comment by StochasticEntropy** [comment]
Sounds good, I'll give it a try.

**Comment by StochasticEntropy** [comment]
I just think one of the problems was that the DREAM implementation is publicly available, in a totally independent framework developed by Steinberger, and now here comes google with a paper that tells us all it's better than dream, but google doesn't have the manpower/interest to show the implementation to the world. So I assume that it's not as good as DREAM. And given that there is no follow up paper I assume that and I'm surprised that there's no follow up improvement.

**Comment by StochasticEntropy** [comment]
On a quiet weekend I might even have a look at the math which seems a bit too optimistic for me, but I trust a bunch of PhDs there for now ;)

**Comment by StochasticEntropy** [comment]
Maybe if you could open a private branch for that, I would appreciate it

**Comment by StochasticEntropy** [comment]
I'm also not a big fan of the model-free paradigm. I think it will always be helpful to train a model to give you some probabilities. Even if we don't have a simulator, we can try to learn the environment and then replay it.

**Comment by lanctot** [comment]
> Maybe if you could open a private branch for that, I would appreciate it

You can use a PR for that, it acts in every way like your own private branch. Don't worry about opening a PR even before it is ready; you can put it in draft mode. And then I can easily go through the code and ask about it. Depending on how complete it is, we could even pull it in early and mark it as experimental so that I can modify it too. We did this recently with C++ DQN and it worked quite well.

**Comment by StochasticEntropy** [comment]
I was more thinking a private conversation branch, believe it or not, I might be a nice person ;)

**Comment by StochasticEntropy** [comment]
I'll do it, but I don't want every bit of conversion in the open.

**Comment by lanctot** [comment]
Got it, but the code being open is ok? You can also reach me/us by email too for the discussion part. 

Heads-up. I'm giving a talk tomorrow which will include an quick overview  (maybe 5 minutes worth) of ARMAC: https://sites.google.com/view/comarl-seminars/upcoming-seminars. It's free to attend but will be recorded in case you miss it.

**Comment by StochasticEntropy** [comment]
Yes, of course the final code is open. That's what were here for.

**Comment by StochasticEntropy** [comment]
I won't miss your talk ;P

**Comment by StochasticEntropy** [comment]
let's see, I would just take the DeepCFR pipeline and integrate the steps accordingly

**Comment by StochasticEntropy** [comment]
as a first step, and then we only have to fix the policies around it

**Comment by lanctot** [comment]
@StochasticEntropy just curious if you're currently working on this? I'm thinking about looking into it with the help of @jhtschultz. Even if it's only partly completed, feel free to submit early because then maybe we can iterate. If you haven't had a chance to start, then maybe we can start it and you can help us too.

**Comment by StochasticEntropy** [comment]
Hi @lanctot, I didn't have a chance to do much yet, but I should have more time in August to help. Do you know if there is already a version of the Tree-Backup implemented somewhere in Open Spiel?

**Comment by StochasticEntropy** [comment]
Generally from my understanding, we could just use most of the DeepCFR code for data gathering with the policy selection modification. However I think the main new thing that has to be implemented is the critic training. The cited TreeBackup Paper is for the original tabular algorithm, so to get that one right here for the neural network training will be the tricky part I think.

**Comment by lanctot** [comment]
Ah apologies, just realized I had not replied to you. Yeah we'd have to implement Tree Backup, but I think that wouldn't be too hard. I'm also not very clear on how important it was for the results and there might be an easier one we can start with like Q(\lambda) or a better one like Retrace. They all look fairly simple to me but I agree they can be tricky to get right. I'll chat about it with the main author.

**Comment by StochasticEntropy** [comment]
Wasn't most of your PhD thesis about importance sampling which is the contrary to Tree Backup ;P

**Comment by lanctot** [comment]
Ahh, yes.. ! But that was also 8-12 years ago! :-p More importantly there was zero function approximation in what I looked at. That was only introduced in RCFR.

ARMAC is entirely about getting rid of the importance sample by using RL-inspired tricks (like off-policy evaluation). Turns out the model-free RL style of sampling is super high variance and you need some tricks to ensure it combines well with function approximation. I think you can get away with it if the games are short (e.g. poker) but for imperfect information games with long episodes, those importance sampling corrections would get too wild. Just my opinion. 

It's definitely not obvious IMO, especially because DREAM was able to get that kind of approach to scale up to a large Poker subgame (using variance reduction baselines whereas our MC-RCFR baseline didn't). I think the trick it uses to normalize the values is quite similar to the ARMAC trick (in practice), so I wonder how important that part was. Either way, we'll need further research to figure that out which is why I think an open general implementation is worth looking into.

**Comment by StochasticEntropy** [comment]
@lanctot calls poker an easy small and short game ;D

**Comment by lanctot** [comment]
Haha .. oh, that was a typo -- sorry! 

What I meant was small (like Leduc, Liar's dice) *or* short (poker). The poker domains, even subgames or action-abstracted games like FCPA.  are _definitely_ not easy :)

**Comment by StochasticEntropy** [comment]
It's really hard to understand how importance sampling works in NN approximations, they should normally already account for importance sampling if the exploration and data gathering is done right. My brain is running in circles when I'm thinking about it and I neither like VR-CFR nor DREAM when they involve NNs from their argumentation.

**Comment by StochasticEntropy** [comment]
Same problem with converting tabular Tree Backup to NNs, because the statistics are messed up when the NN already learned values from nearby states.

**Comment by StochasticEntropy** [comment]
Probably the answer will be evolutionary non-importance-sampling algorithms

**Comment by StochasticEntropy** [comment]
Anyways, I can provide a template based on the DeepCFR code, where the critic training and policy selection is blank

**Comment by StochasticEntropy** [comment]
The more I think about is the more I get confused, I know the paper says that the W think is like the cumulative Regret, and there is a proof in the Appendix, I'm assuming that is also correct for approximations?

**Comment by StochasticEntropy** [comment]
Did Sutton ever say that these NN approximations are theoretically sound?

**Comment by dvstter** [comment]
> our impl is looking? Would you be willing to contrib

Sorry, I had been wondering about other questions for a few months. I have returned back to this question. I am reviewing my code and the algorithm described in the paper. Within this month, I will do some progress, I guess. If there is any progress on this topic by others, please notify me, thanks.

**Comment by dvstter** [comment]
> @dvstter just curious how your impl is looking? Would you be willing to contribute it to OpenSpile?

Sorry, I had been wondering about other questions for a few months. I have returned back to this question. I am reviewing my code and the algorithm described in the paper. Within this month, I will do some progress, I guess. If there is any progress on this topic by others, please notify me, thanks.

**Comment by StochasticEntropy** [comment]
Sorry, I'm also swamped with other projects currently and I expected this one to be easier, but you can always put me down for review.

**Comment by dvstter** [comment]
<img width="717" alt="Screen Shot 2021-09-13 at 4 39 31 PM" src="https://user-images.githubusercontent.com/11886450/133052124-7b31699d-6bf2-4598-8371-85dc71fc3a36.png">
I have three questions about this algorithm, @lanctot @StochasticEntropy could you please spare your time to help me?
1. orange part: I have already discussed with you, these two lines' code seems to be futile;
2. red part: the network q's parameter is \theta^jj, but the network is the critic part for this algorithm which in my opinion, should be updated continually. besides, this algorithm tells that it will use past policy nor a past critic. so I think this should be revised to \theta^t;
3. yellow part: (h, s, a, r, \pi) recorded in the d must satisfied 'where player i acts', so in the training part, the \tau(s) should always be i, consequently, the average policy will never be trained.

**Comment by lanctot** [comment]
Hi @dvstter,

1. I would not call them "futile". The policy is required for several steps of the algorithm and the critic is required to compute the regret, so they are just being defined up here for completion. The first one could have also been expresses as "let \pi^t(s) = NormalizedRelu(\bar{W}_{\theta^t}(s)) **for all timesteps t**"

2. Sorry, I don't understand the question. Can you rephrase? That regret should definitely not be with respect to \theta^t. Computing the W properly requires you to collect regrets from the previous epochs, that is what I called the "ARMAC trick" in my [COMARL presentation here](https://www.youtube.com/watch?v=ycQqCT7FFRw&t=2430s). That is also why we chose the name "retrospective replays": to train \bar{W}_{\theta^t} you need regrets from previous epochs.

3. Ah yes, good catch. This is a typo. You should also be adding data from the histories where the opponents act because it is used to predict the average policy \bar{\pi}. If I were to fix the pseudo-code, I would add this as a separate case:

For history h in \rho where player i acts: 
   // Same as what is there now

For history h in \rho where player i does not act:
   append (h, s, a, null, \pi^j(s))

Hope this helps!

**Comment by dvstter** [comment]
> 2\. we use both past policies (\pi^j, where j is from a past epoch) _and_ past critics,

Many thanks, helps a alot,  but I have one more question.
As you said, armac will use past policies and past critic, so in fact, this algorithm need to implement 4 networks for W and q(current and past), right? and the retrospective buffer will need to save two types of buffer, which namely past policies buffer and past q buffer. Is that right?

**Comment by lanctot** [comment]
I believe we used 4 heads on the same single network. @agruslys, can you confirm precisely what the heads were?

I believe only one replay buffer, I will let Audrunas follow-up with more detail.

**Comment by agruslys** [comment]
The algorithm needs to implement the following networks or heads of the same network:
Q_t(h, a) network, which is a global critic and takes observations of all players as input. It evaluates some past policy pi_t at epoch t where 0 < t <= T.
W_t(s, a) network that takes observations only of the current player. Policy pi_t is defined by normalising max(W_t(s, .), 0), so whether we call it a separate network is a question of semantics.
mu_t(s, a) is the mean policy - this is what converges to Nash.

here 0 < t < T where T is the current epoch, and t refers to the previous epochs. 

The retrospective buffer (that should ideally contain all past epoch networks) will have to save Q_t and W_t (and thus pi_t)

**Comment by dvstter** [comment]
> The algorithm needs to implement the following networks or heads of the same network:
> Q_t(h, a) network, which is a global critic and takes observations of all players as input. It evaluates some past policy pi_t at epoch t where 0 < t <= T.
> W_t(s, a) network that takes observations only of the current player. Policy pi_t is defined by normalising max(W_t(s, .), 0), so whether we call it a separate network is a question of semantics.
> mu_t(s, a) is the mean policy - this is what converges to Nash.
> 
> here 0 < t < T where T is the current epoch, and t refers to the previous epochs.
> 
> The retrospective buffer (that should ideally contain all past epoch networks) will have to save Q_t and W_t (and thus pi_t)

Thanks for your reply. 
So in my opinion, there are four networks from another perspective, which namely past-Q_t, past W_t, Q_t, W_t, mu_t, but the two past-* networks share the same backbone layers with Q_t and W_t. Is that right?

**Comment by lanctot** [comment]
I think the easiest way to see it is as follows. Every network has three output heads as described by Audrunas.

Suppose T = 100. Then, there would be 101 networks in total (since 100 epochs have passe, and for the current epoch), one for t = 0, 1, 2, ..., 100. 

So when t = 100, you are generating data for and training the *current network* at epoch t only. All the past 100 networks (t = 0, 1, 2... 99) are _frozen and never change anymore_, however you use them to generate data to train the network at iteration t. 

What Audrunas meant by \mu_t is actually \bar{\pi}_t in the paper.  (\mu is actually the behavior policy, which is a mixture over other policies plus exploration, not a network head by itself)

**Comment by dvstter** [comment]
> I think the easiest way to see it is as follows. Every network has three output heads as described by Audrunas.
> 
> Suppose T = 100. Then, there would be 101 networks in total (since 100 epochs have passe, and for the current epoch), one for t = 0, 1, 2, ..., 100.
> 
> So when t = 100, you are generating data for and training the _current network_ at epoch t only. All the past 100 networks (t = 0, 1, 2... 99) are _frozen and never change anymore_, however you use them to generate data to train the network at iteration t.
> 
> What Audrunas meant by \mu_t is actually \bar{\pi}_t in the paper. (\mu is actually the behavior policy, which is a mixture over other policies plus exploration, not a network head by itself)

That's amazing. Thanks.

**Comment by dvstter** [comment]
Hello everybody, I just wrote down this code based on my understanding of this paper. 
The link is here [https://github.com/dvstter/cfr-algos/blob/master/armac.py](armac-github-blob).
In fact, I am not sure about this implementation's correctness, so I hope somebody can help me review this code when he is not busy. I am running this code on my server, but it will consume a lot of time. 

There are two important points about this implementation: 
1. I use the simplest Q-learning to replace the Treebackup. Because that is not quite so hard to modify, the foremost important question is that whether this code is right or not. 
2. I was just curious about the 'adaptive policy selection', so my mind is still absorbing in that paper.

**Comment by lanctot** [comment]
I think it's a great idea to try Q-learning before Tree Backup and to skip the part about adaptive policy selection for the first version.

It would be a lot easier to review if you were able to submit it as a PR on this repository. Would it be easy to that? Because then we can go back and forth, and we can even import the first version of it (before it's officially ready) if anyone wants to help you continue with it.

**Comment by dvstter** [comment]
> I think it's a great idea to try Q-learning before Tree Backup and to skip the part about adaptive policy selection for the first version.
> 
> It would be a lot easier to review if you were able to submit it as a PR on this repository. Would it be easy to that? Because then we can go back and forth, and we can even import the first version of it (before it's officially ready) if anyone wants to help you continue with it.

Thanks, I just submitted it.

**Comment by lanctot** [comment]
Thanks @dvstter ! Implementation is here: https://github.com/deepmind/open_spiel/pull/706, so I'm now closing this.

**Comment by StochasticEntropy** [comment]
Awesome, sorry that I didn’t have more time to contribute anymore
